libraries/ms-common/src/can.c

```diff
@@ -0,0 +1,155 @@
+
+    StatusCode result = can_queue_push(&can->rx_queue, &rx_msg);
+    if (result != STATUS_CODE_OK) {
+      return;
+    }
+
+    counter++;
+  }
+
+  if (counter != 0) {
+    event_raise(can->rx_event, counter);
+  }
+}
+
+// Bus error timer callback
+// Checks if the bus has recovered, raising the fault event if still off
+void prv_bus_error_timeout_handler(SoftTimerID timer_id, void *context) {
+  CANConfig *can = context;
+
+  if (can_hw_bus_status(&can->hw) == CAN_HW_BUS_STATUS_OFF) {
+    event_raise(can->fault_event, 0);
+  }
+}
+
+// Handler for CAN HW bus error events
+// Starts a timer to check for bus recovery
+void prv_bus_error_handler(void *context) {
+  CANConfig *can = context;
+
+  soft_timer_start_millis(CAN_BUS_OFF_RECOVERY_TIME_MS, prv_bus_error_timeout_handler, can, NULL);
+}
+
+StatusCode can_init(CANConfig *can, uint16_t device_id, uint16_t bus_speed, bool loopback,
+                    EventID rx_event, EventID fault_event) {
+  if (device_id >= CAN_MSG_MAX_DEVICES) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "CAN: Invalid device ID");
+  }
+
+  memset(can, 0, sizeof(*can));
+
+  can_hw_init(&can->hw, bus_speed, loopback);
+
+  can_rx_fsm_init(&can->fsm, can);
+  can_queue_init(&can->tx_queue);
+  can_queue_init(&can->rx_queue);
+  can_ack_init(&can->ack_requests);
+  can_rx_init(&can->rx_handlers, can->rx_handler_storage, CAN_MAX_RX_HANDLERS);
+
+  can->device_id = device_id;
+  can->rx_event = rx_event;
+  can->fault_event = fault_event;
+
+  can_hw_register_callback(&can->hw, CAN_HW_EVENT_TX_READY, prv_tx_handler, can);
+  can_hw_register_callback(&can->hw, CAN_HW_EVENT_MSG_RX, prv_rx_handler, can);
+  can_hw_register_callback(&can->hw, CAN_HW_EVENT_BUS_ERROR, prv_bus_error_handler, can);
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode can_add_filter(CANConfig *can, CANMessageID msg_id) {
+  if (msg_id >= CAN_MSG_MAX_IDS) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "CAN: Invalid message ID");
+  }
+
+  CANId can_id = {
+    .msg_id = msg_id
+  }, mask = {
+    .msg_id = UINT16_MAX
+  };
+
+  return can_hw_add_filter(&can->hw, can_id.raw, mask.raw);
+}
+
+StatusCode can_register_rx_default_handler(CANConfig *can, CANRxHandlerCb handler, void *context) {
+  return can_rx_register_default_handler(&can->rx_handlers, handler, context);
+}
+
+StatusCode can_register_rx_handler(CANConfig *can, CANMessageID msg_id,
+                                   CANRxHandlerCb handler, void *context) {
+  return can_rx_register_handler(&can->rx_handlers, msg_id, handler, context);
+}
+
+StatusCode can_transmit(CANConfig *can, const CANMessage *msg, const CANAckRequest *ack_request) {
+  if (msg->msg_id >= CAN_MSG_MAX_IDS) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "CAN: Invalid message ID");
+  }
+
+  if (ack_request != NULL) {
+    if (!CAN_MSG_IS_CRITICAL(msg)) {
+      return status_msg(STATUS_CODE_INVALID_ARGS, "CAN: ACK requested for non-critical message");
+    }
+
+    StatusCode ret = STATUS_CODE_OK;
```

Comment:
Personally, I'm okay with the long line as it took me longer to figure out the point of this than the long line would have, but up to you? If you stick with this I would consider not assigning `STATUS_CODE_OK` as the function call is just going to update `ret` and that assignment confused me the most.